### PR TITLE
HDDS-9117. Source bucket replication type is ignored for links

### PR DESCRIPTION
**Pull request by Cyrill**
## What changes were proposed in this pull request?

Before these changes single and multipart uploads via S3G to an EC bucket resulted in keys created with RATIS/3 replication. 
Now linked buckets inherit the original bucket's replication type

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9117

## How was this patch tested?

Unit tests, manual tests

### CODE REVIEW

**Comment by umamaheswararao** [comment]
@SaketaChalamchala could you please help with this?

**Comment by SaketaChalamchala** [comment]
Thanks for the patch @Cyrill. 
It looks like the problem here is that keys are getting created with server default RATIS 3 replication because linked buckets are not inheriting the source bucket's replication config. The fix for linked buckets to inherit source bucket's replication config should suffice to resolve this. All keys created under the linked bucket should inherit the replication config from the bucket args.
Was there a reason to modify S3G multipart upload method to accept replication config?

**Review by errose28 — COMMENTED** [review]
Thanks for looking at this @Cyrill. While the change does work, I'm not sure it is the best way to fix the problem.

The method `OzoneManagerUtils#resolveLinkBucketLayout` is only intended to resolve the bucket layout, not other properties like rep config or quota. Quota has a similar problem, where it can be set on the source but not returned in the bucket link's info. We need a more general solution for getting the properties of the source bucket when working with bucket links. The only real caller of this method is `BucketManagerImpl#getBucketInfo`, which can be called by `OzoneManager#resolveBucketLink`, meaning the bucket links can be resolved twice.

I think we should get rid of this method and make sure bucket resolution is happening only once through a proper channel like `OzoneManager#resolveBucketLink`. A unit test that retrieves bucket info for a link and source and checks that only the names are different can validate this for all pieces of bucket metadata.

**Comment by kerneltime** [comment]
> Thanks for looking at this @Cyrill. While the change does work, I'm not sure it is the best way to fix the problem.
> 
> The method `OzoneManagerUtils#resolveLinkBucketLayout` is only intended to resolve the bucket layout, not other properties like rep config or quota. Quota has a similar problem, where it can be set on the source but not returned in the bucket link's info. We need a more general solution for getting the properties of the source bucket when working with bucket links. The only real caller of this method is `BucketManagerImpl#getBucketInfo`, which can be called by `OzoneManager#resolveBucketLink`, meaning the bucket links can be resolved twice.
> 
> I think we should get rid of this method and make sure bucket resolution is happening only once through a proper channel like `OzoneManager#resolveBucketLink`. A unit test that retrieves bucket info for a link and source and checks that only the names are different can validate this for all pieces of bucket metadata.

`getBucketInfo` can be implemented to first resolve the bucket at the end of the nest chain link and then call the `metadataManager` to get the `OmBucketInfo`. That should preserve the behavior of the call `getBucketInfo`  independent of the linking.

**Comment by kerneltime** [comment]
Can you take a look at the failed CI runs.

For uploading an object to a bucket where the source bucket does not exist, it should error out. Currently the code does not prevent dangling references but writing objects should fail.

**Review by errose28 — COMMENTED** [review]
Thanks for the updates @Cyrill. I agree there are more issues with links than we can fix in one PR. Since it sounds like you are aware of the some of the gaps, could you file a parent jira with subtasks to address the remaining issues, and better define the scope of this one jira? We can have others work on the remaining tasks if you do not have the time, but it would be good to capture what is left after this task.

**Comment by krishnaasawa1** [comment]
@ChenSammi @ashishkumar50 can you please help with review

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -2792,7 +2792,32 @@ public OmBucketInfo getBucketInfo(String volume, String bucket)
             bucket, null);
       }
       metrics.incNumBucketInfos();
-      return bucketManager.getBucketInfo(volume, bucket);
+
+      ResolvedBucket resolvedBucket =
+          resolveBucketLink(Pair.of(volume, bucket), true);
```
Comment (ashishkumar50) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2797]: Even bucket is not `linkBucket` still we are resolving bucket, that looks like is not necessary. I believe all these operations which is currently been done inside `resolvedBucketLink` is required only for link buckets. In the old code it just skips doing for link buckets.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 2]
```diff
@@ -898,6 +898,45 @@ public void testDeleteBucket()
...
+    //now linkedBucketName has become a dangling one
+    //should still be possible to get its info
+    OzoneBucket danglingLinkedBucket = volume.getBucket(linkedBucketName);
+    assertNotNull(danglingLinkedBucket);
+
+    //now delete the dangling linked bucket
+    volume.deleteBucket(linkedBucketName);
+
+    OzoneTestUtils.expectOmException(ResultCodes.BUCKET_NOT_FOUND,
+        () -> volume.getBucket(bucketName)
```
Comment (ashishkumar50) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:933]: Better this check we can move after L923 to know source bucket is deleted and link still exist.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 3]
```diff
@@ -898,6 +898,45 @@ public void testDeleteBucket()
...
+    );
+    OzoneTestUtils.expectOmException(ResultCodes.BUCKET_NOT_FOUND,
+        () -> volume.getBucket(linkedBucketName)
+    );
```
Comment (ashishkumar50) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:937]: Delete created `volume` at the end.

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
@Cyrill, we can merge this soon, I was just waiting to see if other reviewers (especially @ashishkumar50 as most recent one, but also @errose28, @kerneltime, @SaketaChalamchala) want to have another look.

**Comment by errose28** [comment]
I will likely not get time to come back to this. No need to block the PR on my review.

**Comment by ashishkumar50** [comment]
@Cyrill Thanks for updating patch, LGTM+1.

**Comment by adoroszlai** [comment]
Thanks @Cyrill for the fix, @ashishkumar50, @errose28, @kerneltime, @SaketaChalamchala for the review.

**Pull request merged**